### PR TITLE
SharedTerm rewrite

### PR DESCRIPTION
**Pull request by jhoenicke**
Get rid of SharedTerm, use Term instead. Code is tested with both quantifier free and some quantified benchmarks.

TODO: Use the ClausifierTermInfo also for Literals and simplify the code.

### CODE REVIEW

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java [hunk 1]
```diff
@@ -1226,40 +1219,123 @@ public SharedTerm getSharedTerm(final Term t, final boolean inCCTermBuilder, fin
...
-					addUnshareLA(res);
+					final MutableAffineTerm mat = createMutableAffinTerm(new SMTAffineTerm(term), source);
+					assert mat.getConstant().mEps == 0;
+					shareLATerm(term, termInfo, new LASharedTerm(term, mat.getSummands(), mat.getConstant().mReal));
 				}
 			}
 		}
+	}
+
+	public LinVar getLinVar(final Term term) {
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java]: Method only for basic term, right?

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/theory/cclosure/CClosure.java [hunk 2]
```diff
@@ -938,25 +932,27 @@ public void dumpModel(final LogProxy logger) {
...
 	private boolean checkCongruence() {
 		boolean skip;
 		for (final CCTerm t1 : mAllTerms) {
+			if (!(t1 instanceof CCAppTerm)) {
+				continue;
+			}
+			final CCAppTerm a1 = (CCAppTerm) t1;
 			skip = true;
 			for (final CCTerm t2 : mAllTerms) {
+				// don't check symmetric cases: we skip all terms until we find the first term.
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/theory/cclosure/CClosure.java]: Please edit this comment.

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/theory/cclosure/CClosure.java [hunk 3]
```diff
@@ -51,39 +52,94 @@
...
 import de.uni_freiburg.informatik.ultimate.util.datastructures.ScopedHashMap;
 
+/**
+ * This class implements the theory of equality, a.k.a. congruence closure.
+ *
+ * This theory basically handles equality literals in particular CCEquality. It does also Nelson-Oppen propagation For
+ * every subterm in the formula that is handled by this theory there is a corresponding CCTerm object. These in the
+ * sense that it sets the LAEquality if a corresponding CCEquality was set. CCTerm object build the basic union-merge
+ * datastructure.
+ *
```
Comment (tanjaschindler) [inline SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/theory/cclosure/CClosure.java]: Please edit this comment, it is a bit unclear.

File: SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java (hunk 1 above)
Comment (jhoenicke) [reply SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/convert/Clausifier.java]: may be only called for "initiallynonbasic term", i.e. a term that can appear in the summand hashmap of an SMTAffineTerm. I can add an assertion like
`term == new SMTAffineTerm(term).getSummands().keySet().iterator().next()`

Currently this is only used by model production, to get the value for each summand for a congruence-class that has a shared term.

**Pull request merged**